app/ui/browser-modern/views/browser/page/index.jsx

```diff
@@ -97,26 +99,66 @@ class Page extends Component {
-      // Not emitted when a failure occurs, unlike `did-stop-loading`.
       // Might not always be emitted after a `did-start-loading`.
     });
 
-    this.webview.addEventListener('did-fail-load', () => {
-      // Like 'did-finish-load', but fired when the load failed or was cancelled.
-      // May be emitted multiple times for every single frame.
-      // Only emitted when a failure occurs, unlike `did-stop-loading`.
-      this.props.dispatch(PageActions.setPageState(this.props.pageId, {
-        load: PageState.STATES.FAILED,
-      }));
-    });
-
     this.webview.addEventListener('did-stop-loading', () => {
       // Event fired when a page finishes loading.
       // May be emitted multiple times for every single frame.
       // Emitted regardless of whether or not a failure occurs.
-      this.props.dispatch(PageActions.setPageState(this.props.pageId, {
-        load: PageState.STATES.LOADED,
+      const title = this.webview.getTitle();
+      const url = this.webview.getURL();
+
+      // `page-title-set` is not called on error pages, nor is
+      // `did-navigate` called, so we must set these properties to
+      // properly render our error pages
+      this.props.dispatch(PageActions.setPageDetails(this.props.pageId, {
+        title,
+        location: url,
       }));
+
+      // If the page state is still LOADING, and we haven't hit a
+      // failure state, mark this page as LOADED.
+      if (this.props.pageState.load === PageState.STATES.LOADING) {
+        this.props.dispatch(PageActions.setPageState(this.props.pageId, {
+          load: PageState.STATES.LOADED,
+        }));
+      }
+    });
+
+    this.webview.addEventListener('did-fail-load', (e) => {
+      // Like 'did-finish-load', but fired when the load failed or was cancelled.
+      // May be emitted multiple times for every single frame.
+      // Only emitted when a failure occurs, unlike `did-stop-loading`.
+      // Is fired before did-finish-load and did-stop-loading when failure
+      // occurs.
+      const { errorCode, errorDescription, validatedURL, isMainFrame } = e;
+
+      if (!isMainFrame) {
+        return;
+      }
+
+      // If a page is aborted (like hitting BACK/RELOAD while a page is loading), we'll get a
+      // load failure of ERR_ABORTED with code `-3` -- this is intended, so just ignore it.
+      if (Math.abs(errorCode) === 3) {
+        return;
+      }
+
+      const state = {
+        load: PageState.STATES.FAILED,
+        code: errorCode,
+        description: errorDescription,
+      };
+
+      // Most cert errors are 501 from this event; the true, more descriptive
+      // error description can be retrieved from the main process. Just check
+      // here roughly if the error looks like a cert error.
+      if (Math.abs(errorCode) === 501 || /CERT/.test(errorDescription)) {
+        this.props.dispatch(PageEffects.getCertificateError(this.props.pageId, validatedURL));
+      }
```

Comment:
There's lots and lots of difficult to test logic in here. I'd move most of it into utility functions in some module.